### PR TITLE
fix(amplitude): added an error validation at processor

### PR DESCRIPTION
**Pull request by sandeepdsvs**
## Description of the change

> Added an error validation at processor, throwing error if both `user_id` and `device_id` not present.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/am/transform.js [hunk 1]
```diff
@@ -499,6 +499,23 @@ function responseBuilderSimple(
           delete payload.user_properties.address.country;
         }
       }
+
+      if (!payload.user_id && !payload.device_id) {
+        logger.debug("Both userId and deviceId cannot be undefined");
+        throw new ErrorBuilder()
```
Comment (sanpj2292) [inline src/v0/destinations/am/transform.js:505]: Can we change to new error class convention, confirm what error class does this kind of error belong to & make relevant changes ?

File: src/v0/destinations/am/transform.js [hunk 2]
```diff
@@ -499,6 +499,23 @@ function responseBuilderSimple(
...
+          .setStatus(400)
+          .setMessage("Both userId and deviceId cannot be undefined")
```
Comment (saikumarrs) [inline src/v0/destinations/am/transform.js:507]:
```suggestion
          .setMessage("Either of user ID or device ID fields must be specified")
```

**Review by saikumarrs — CHANGES_REQUESTED** [review]

**Review by ItsSudip — APPROVED** [review]

**Comment by sanpj2292** [comment]
@sandeepdsvs If this change takes more time, you can take it up later. But we should also validate the stat-tags that are generated when we throw an error in test-cases. We can use `expect().toThrow()`(on those lines), I think to validate so that we have complete clarity when look at test-cases.

File: src/v0/destinations/am/transform.js [hunk 3]
```diff
@@ -499,6 +499,14 @@ function responseBuilderSimple(
           delete payload.user_properties.address.country;
         }
       }
+
+      if (!payload.user_id && !payload.device_id) {
+        logger.debug("Either of user ID or device ID fields must be specified");
```
Comment (sanpj2292) [inline src/v0/destinations/am/transform.js:504]: [nitpick] Would a debug-log add value here ?

**Review by sanpj2292 — APPROVED** [review]

File: src/v0/destinations/am/transform.js (hunk 3 above)
Comment (sanpj2292) [reply src/v0/destinations/am/transform.js:504]: This is a non-blocking comment

**Review by saikumarrs — APPROVED** [review]

File: src/v0/destinations/am/transform.js (hunk 3 above)
Comment (sandeepdsvs) [reply src/v0/destinations/am/transform.js:504]: Followed the existing convention in the amplitude destination, but later realised it's outdated. I will once go through the amplitude doc and accordingly will try to raise a refactor/ new V1 version transformer task, where I'll try to cover all things. For now will merge this

**Pull request merged**